core/src/main/java/hudson/model/Job.java

```diff
@@ -684,9 +685,37 @@ public void movedTo(DirectlyModifiableTopLevelItemGroup destination, AbstractIte
         }
     }
 
-    @Override public void delete() throws IOException, InterruptedException {
+    @Override
+    public void delete() throws IOException, InterruptedException {
         super.delete();
         Util.deleteRecursive(getBuildDir());
+        if (this instanceof TopLevelItem) {
+            cleanupWorkspaces((TopLevelItem) this);
+        }
+    }
+
+    /**
+     * Clean out workspaces for attached agents (best efforts, as they may not always be online)
+     * There is also hudson.model.WorkspaceCleanupThread to catch ones that were missed.
+     */
+    private void cleanupWorkspaces(TopLevelItem item) {
+        Jenkins j = Jenkins.getInstance();
+        List<Node> nodes = new ArrayList<>();
+        nodes.add(j);
+        nodes.addAll(j.getNodes());
+        Logger log = Logger.getLogger(Job.class.getName());
+        for (Node node : nodes) {
+            FilePath ws = node.getWorkspaceFor(item);
```

Comment:
my main area of concern with this is how it scales when you have a very large number of nodes.

this will block the web UI thread - and slaves may have slow disks (aka spinning rust and have lots of small files) so could take an amount of time.
Perhaps it may be better to kick the workspace cleanup thread instead and handle this asynchronously?